fix ionization

BrianMarre opened the pull request:
fix for [Issue #4611](https://github.com/ComputationalRadiationPhysics/picongpu/issues/4611)

bug introduced with #4592

BrianMarre committed:
fix ionization kernels due to changed SetToAtomicGroundStateForChargeState

BrianMarre commented:
@psychocoderHPC ready for review

psychocoderHPC commented on include/picongpu/particles/ionization/byCollision/ThomasFermi/ThomasFermi_Impl.hpp:
```diff
@@ -309,15 +309,20 @@ namespace picongpu
+                    parentIon[boundElectrons_] -= float_X(1.);
+
+                    uint8_t const newNumberBoundElectrons
+                        = static_cast<uint8_t>(parentIon[boundElectrons_]) - static_cast<uint8_t>(1u);
```
we should avoid readin the value `parentIon[boundElectrons_]` twice. IMO this change is also reducing the bound electrons twice.

psychocoderHPC replied:
offline discussed with @BrianMarre I am correct that we ionize now twice.
He will introduce a new method to set the charge state where bound electrons and atomic physics ground states will be handled correctly.

psychocoderHPC replied:
After the fix there should be no code in PIConGPU except in the new method where `parentIon[boundElectrons_] = ...` is called.

BrianMarre commented:
One further note, we inadvertently break all setups using the `SetToAtomicGroundStateForChargeState` functor for charge state initialization.

Look for something like the following in your setup:
```C++
[particle.param]
namespace picongpu::manipulators
{
    /** Define initial ion charge */
    struct SetIonTo16Ionized
    {
        /** set ion bound electrons to yield targetChargeState.
         */
        template<typename T_Particle>
        DINLINE void operator()(T_Particle& particle)
        {
            constexpr float_X targetChargeState = 16;

            // number of Electrons for neutral atom
            constexpr float_X numberBoundElectrons
                = picongpu::traits::GetAtomicNumbers<T_Particle>::type::numberOfProtons;

            PMACC_ASSERT_MSG(
                numberBoundElectrons >= targetChargeState,
                "too few protons for target charge state");

            picongpu::particles::atomicPhysics::SetToAtomicGroundStateForChargeState{}(
                particle,
                numberBoundElectrons - targetChargeState);
        }
    };

    /** Definition of manipulator that sets properties of an ion macro-particle
     * according properties defined in struct SetIonTo16Ionized.
     */
    using SetIonization = generic::Free<SetIonTo16Ionized>;
} // namespace picongpu::particles::manipulators
```
There is nothing I can do about that.

The init-functor must either be replaced entirely with
```c++
namespace picongpu::particles::manipulators
{
    using SetIonization = unary::ChargeState<16u>;
} // namespace picongpu::particles::manipulators
```
or
`SetToAtomicGroundStateForChargeState` must be replaced with `SetChargeState`

BrianMarre commented:
@psychocoderHPC CI failed with
```bash
Running with gitlab-runner 16.0.1-hifis0 (14b60006)
  on hifis-runner-manager-gitlabcom-1 9kYz4ey9, system ID: s_b907b9cd0eff
Resolving secrets 00:00
Preparing the "docker+machine" executor 04:06
Using Docker executor with image registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04-rocm5.2-pic:3.0 ...
Pulling docker image registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04-rocm5.2-pic:3.0 ...
Using docker image sha256:c4f8545301b5acabbbae453b112f3ab9e8ed0dee3cb9bf8a73780722a0ddca0c for registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04-rocm5.2-pic:3.0 with digest registry.hzdr.de/crp/alpaka-group-container/alpaka-ci-ubuntu20.04-rocm5.2-pic@sha256:ec18e6cff78daf0071055b25d6278ddc603ba23493145eb588d113f35f1f8038 ...
Preparing environment 00:01
Running on runner-9kyz4ey9-project-19922452-concurrent-0 via runner-9kyz4ey9-hifis-autoscale-1687355179-7ca5afc1...
Getting source from Git repository 00:01
Fetching changes with git depth set to 50...
Initialized empty Git repository in /builds/hzdr/crp/picongpu/.git/
Created fresh repository.
fatal: unable to access 'https://gitlab.com/hzdr/crp/picongpu.git/': OpenSSL/3.1.0: error:0A000410:SSL routines::sslv3 alert handshake failure
Cleaning up project directory and file based variables 00:00
ERROR: Job failed: exit code 1
```

psychocoderHPC commented:
> One further note, we inadvertently break all setup using the SetToAtomicGroundStateFoChargeState functor for charge state initialization
> 
> ```c++
> [particle.param]
> namespace picongpu::manipulators
> {
>     /** Define initial ion charge */
>     struct SetIonTo16Ionized
>     {
>         /** set ion bound electrons to yield targetChargeState.
>          */
>         template<typename T_Particle>
>         DINLINE void operator()(T_Particle& particle)
>         {
>             constexpr float_X targetChargeState = 16;
> 
>             // number of Electrons for neutral atom
>             constexpr float_X numberBoundElectrons
>                 = picongpu::traits::GetAtomicNumbers<T_Particle>::type::numberOfProtons;
> 
>             PMACC_ASSERT_MSG(
>                 numberBoundElectrons >= targetChargeState,
>                 "too few protons for target charge state");
> 
>             picongpu::particles::atomicPhysics::SetToAtomicGroundStateForChargeState{}(
>                 particle,
>                 numberBoundElectrons - targetChargeState);
>         }
>     };
> 
>     /** Definition of manipulator that sets properties of an ion macro-particle
>      * according properties defined in struct SetIonTo16Ionized.
>      */
>     using SetIonization = generic::Free<SetIonTo16Ionized>;
> } // namespace picongpu::particles::manipulators
> ```
> 
> and there is nothing I can do about that.
> 
> The init-functor must either be replaced entirely with
> 
> ```c++
> namespace picongpu::particles::manipulators
> {
>     using SetIonization = unary::ChargeState<16u>;
> } // namespace picongpu::particles::manipulators
> ```
> 
> or `SetToAtomicGroundStateForChargeState` must be replaced with `SetChargeState`

Could you explain how old functors using `SetToAtomicGroundStateForChargeState` are broken? If so please check all usage of `SetToAtomicGroundStateForChargeState` everywhere, there are some usages in the param files.

psychocoderHPC commented on include/picongpu/particles/ionization/byField/BSI/BSI_Impl.hpp:
```diff
@@ -26,7 +26,7 @@
 #include "picongpu/fields/FieldJ.hpp"
 #include "picongpu/particles/ParticlesFunctors.hpp"
-#include "picongpu/particles/atomicPhysics/SetToAtomicGroundStateForChargeState.hpp"
+#include "picongpu/particles/atomicPhysics/SetChargeState.hpp"
```
please us ethe include in all param files too or remove them if `SetToAtomicGroundStateForChargeState` is not used there.
in test and there the collision test `SetToAtomicGroundStateForChargeState` is used and must be substituted by `SetChargeState` too.

BrianMarre commented:
In [PR#4592](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4592) I moved the initialization of the macro-ion attribute `boundElectrons` from the `SetToAtomicGroundStateForChargeState` functor to the `ChargeState` Manipulator.

I did not recognize at the time that this broke both the ionization-algorithm and some user setups, as they relied on `SetToAtomicGroundStateForChargeState` to set/init the `boundElectrons`macro-ion attribute.

Not all user setups are affected, since older setups usually set the `boundElectrons` directly in a param defined functor, and are therefore not impacted by the change of `SetToAtomicGroundStateForChargeState`.

Only setups created from the `atomicPhysics` example since [PR#4134](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4134) until [PR#4592](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4592) are affected, as well as all setups created from `CollisionsBeamRelaxation` and `CollisionsThermalisation` from [PR#4418](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4418) until [PR#4592](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4592), as well as all custom setups using `SetToAtomicGroundStateForChargeState` are affected.

With this PR, the ionization algorithms and `ChargeState` manipulator are now using the new SetChargeState functor to set the `boundElectrons`macro-ion attribute, but setups using the intermediate param defined functor for `boundElectron` initialization using `SetToAtomicGroundStateForChargeState` do not actually initialize the `boundElectron` macro-ion attribute, as they do not set it themselves, and rely on `SetToAtomicGroundStateForChargeState` to set it for them.

Since [PR#4592](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4592) also updated all example setups to the new `ChargeState` manipulator, every setup built upon an example since then are fine.

BrianMarre commented:
@psychocoderHPC CI seems to have failed due to outside influences

psychocoderHPC commented on include/picongpu/particles/atomicPhysics/SetChargeState.hpp:
```diff
@@ -0,0 +1,42 @@
+        template<typename T_Ion>
+        DINLINE void operator()(T_Ion& ion, float_X numberBoundElectrons)
+        {
+            PMACC_DEVICE_ASSERT_MSG(numberBoundElectrons >= 0._X, "Number of bound electrons must be >= 0");
```
If we assume that `picongpu::traits::GetAtomicNumbers<T_Particle>::type::numberOfProtons;` is defined for any ion species we could check that `numberBoundElectrons` is not too large.

@BrianMarre doe sit make sense?

psychocoderHPC replied:
offline discussed: IOt makes sense but should go into a seperate PR.

pordyna commented:
@psychocoderHPC can we merge it as it is, so that we don't have the bug on dev anymore? We can add the extra assert in a separate PR, so that we don't need to wait for the CI to run again.
